pkg/notifiers/sender.go

```diff
@@ -0,0 +1,302 @@
+	if notifier.Spec.SlackConfig != nil {
+		if recipient == "" {
+			recipient = notifier.Spec.SlackConfig.DefaultRecipient
+		}
+		return TestSlack(notifier.Spec.SlackConfig.URL, recipient, msg.Content)
+	}
+
+	if notifier.Spec.SMTPConfig != nil {
+		s := notifier.Spec.SMTPConfig
+		if recipient == "" {
+			recipient = s.DefaultRecipient
+		}
+		return TestEmail(s.Host, s.Password, s.Username, int(s.Port), s.TLS, msg.Title, msg.Content, recipient, s.Sender)
+	}
+
+	if notifier.Spec.PagerdutyConfig != nil {
+		return TestPagerduty(notifier.Spec.PagerdutyConfig.ServiceKey, msg.Content)
+	}
+
+	if notifier.Spec.WebhookConfig != nil {
+		return TestWebhook(notifier.Spec.WebhookConfig.URL, msg.Content)
+	}
+
+	return nil
+}
+
+func TestPagerduty(key, msg string) error {
+	if msg == "" {
+		msg = "Pagerduty setting validated"
+	}
+
+	pd := &pagerDutyMessage{
+		ServiceKey:  key,
+		EventType:   "trigger",
+		IncidentKey: hashKey("key"),
+		Description: msg,
+	}
+
+	url := "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(pd); err != nil {
+		return err
+	}
+	resp, err := http.Post(url, "application/json", &buf)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("http status code is not 200")
+	}
+
+	return nil
+}
+
+func TestWebhook(url, msg string) error {
+	if msg == "" {
+		msg = "Webhook setting validated"
+	}
+	alertList := model.Alerts{
+		&model.Alert{
+			Labels: map[model.LabelName]model.LabelValue{
+				model.LabelName("test_msg"): model.LabelValue(msg),
+			},
+		},
+	}
+
+	alertData, err := json.Marshal(alertList)
+	if err != nil {
+		return err
+	}
+
+	resp, err := http.Post(url, "application/json", bytes.NewBuffer(alertData))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		return fmt.Errorf("http status code is %d, not include in the 2xx success HTTP status codes", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func TestSlack(url, channel, msg string) error {
+	if msg == "" {
+		msg = "Slack setting validated"
+	}
+	req := struct {
+		Text    string `json:"text"`
+		Channel string `json:"channel"`
+	}{}
+
+	req.Text = msg
+	req.Channel = channel
+
+	data, err := json.Marshal(req)
+	if err != nil {
+		return err
+	}
+
+	resp, err := http.Post(url, "application/json", bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("http status code is not 200")
+	}
+
+	res, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if string(res) != "ok" {
+		return fmt.Errorf("http response is not ok")
+	}
+
+	return nil
+}
+
+func TestEmail(host, password, username string, port int, requireTLS bool, title, content, receiver, sender string) error {
+	var c *smtp.Client
```

Comment:
this method is a bit long. Any way we can split it by extracting logical parts into their own methods?